### PR TITLE
Promote fetchLinkSuggestions to non-experimental API

### PR DESCRIPTION
**Pull request by getdave**
## ⚠️ DO NOT MERGE

See https://make.wordpress.org/core/2022/04/25/editor-chat-agenda-27-april-2022/#comment-42966

-------

## What?
This PR updates `__experimentalFetchLinkSuggestions` to `fetchLinkSuggestions` thereby making it non-experimental.

## Why?

The API has been in Core Data for a few years now and it's had time to bed in.

Given that the API has been around for so long there is a good chance 3rd party Plugins will utilise the `__experimental` prefixed version. We should be clear to highlight this in release notes.

## How?

All instances of `__experimentalFetchLinkSuggestions` in Core renamed to remove prefix.

## Testing Instructions

* Check Link UI functions in _all_ editors and you are able to search for things.
* Check Link UI functions in Nav block in _all_ editors and you are able to search for things.
* Check tests pass.

### CODE REVIEW

File: packages/core-data/src/fetch/index.js [hunk 1]
```diff
@@ -1,2 +1,2 @@
-export { default as __experimentalFetchLinkSuggestions } from './__experimental-fetch-link-suggestions';
+export { default as fetchLinkSuggestions } from './__experimental-fetch-link-suggestions';
```
Comment (kevin940726) [inline packages/core-data/src/fetch/index.js]: Let's rename the file name to `fetch-link-suggestions.js` too :)

Comment (kevin940726) [reply packages/core-data/src/fetch/index.js]: Along with the unit test file.

**Review by kevin940726 — COMMENTED** [review]
LGTM 👍

Some file names are not renamed yet though 😆

**Comment by getdave** [comment]
> LGTM 👍
> 
> Some file names are not renamed yet though 😆

@kevin940726 Thanks good spot. I renamed things. Look better?

File: packages/core-data/src/fetch/test/__experimental-fetch-link-suggestions.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
 /**
```
Comment (kevin940726) [inline packages/core-data/src/fetch/test/__experimental-fetch-link-suggestions.js:1]: I think we forgot to delete this file 😅

**Review by kevin940726 — APPROVED** [review]
Looks good! Just a tiny oversight left 👍

File: packages/core-data/src/fetch/test/__experimental-fetch-link-suggestions.js (hunk 2 above)
Comment (getdave) [reply packages/core-data/src/fetch/test/__experimental-fetch-link-suggestions.js:1]: Noooooooo! How did this happen 🤦

**Comment by getdave** [comment]
Given how long this has been around we need to give some time for folks to adapt to the change. I'm going to hold on merging until after WP 6.0 and also going to raise in Core Editor chat.

**Comment by getdave** [comment]
I reached out to a few of the Plugins in the directory that I could find who have used this API. I did this as a _personal courtesy_  and not as part of any official process determined by Gutenberg Core.

I'll give this a few days then merge.

**Comment by Clorith** [comment]
Given that this feature has been bundled with core, and is in use on [~800 000-900 000 active installs based off the WordPress.org repository](https://wpdirectory.net/search/01G24JD124541CPNABQ1C1ZXSX), and an unknown amount of paid plugins or themes, is there any reason why an alias is not introduced to maintain backwards compatibility?

**Comment by getdave** [comment]
> Given that this feature has been bundled with core, and is in use on [~800 000-900 000 active installs based off the WordPress.org repository](https://wpdirectory.net/search/01G24JD124541CPNABQ1C1ZXSX), and an unknown amount of paid plugins or themes, is there any reason why an alias is not introduced to maintain backwards compatibility?

This is a good point. You'll recognise my reticence to merge this is based primarily on concerns along the same lines you have mentioned.
 
In general anything marked `__expermental` is considered something that can change or be removed without notice. However, in certain cases that's clearly going to cause a lot of problems. This is one such case.

I believe there is [a wider discussion happening](https://github.com/WordPress/gutenberg/issues/40316) about how we can avoid situations such as this occuring. However, in this case we are where we are.

Let's see how other @WordPress/gutenberg-core  team folks feel but it might be reasonable to introduce a deprecation notice and possibly retain an alias for `__experimental` during an extended deprecation period.

**Comment by youknowriad** [comment]
Hi there! experimental APIs are not considered stable APIs and are documented as such. It's like `@internal` functions in PHP and these also can be changed/renamed AFAIK. https://developer.wordpress.org/block-editor/contributors/code/coding-guidelines/#experimental-and-unstable-apis

So in that sense, there's no difference between how we approached things in Gutenberg and existing practices pre-gutenberg in Core. 

We also did similar renames a lot of times in the past (a quick search in the PR list can show these) which much impact. 

It is true that plugin and theme authors that want to use internal functions are using these and this won't change in the future regardless of what we do. and it's also not something new with Gutenberg as I said before.

For me, I still think we should continue the current approach, it has proved to be good in reducing the impact, communicating things properly to the third-party developers and allowing us to only expose things as stable when we're ready to commit to them long term in terms of backward compatibility. (outside core internal usage). 

--
That said, I agree with @Clorith that we should consider any "stabilization" separately and avoid a general rule. If this particular API is something that folks are relying on heavily, we could add a quick deprecation. I don't advise against deprecations for all experimental APIs as it would create a lot of bloat in our code base for no big added value for third-party developers.

**Comment by getdave** [comment]
> -- That said, I agree with @Clorith that we should consider any "stabilization" separately and avoid a general rule. If this particular API is something that folks are relying on heavily, we could add a quick deprecation. 

I understand you are saying that we should 

- add a deprecation notice for the `__experimental` API
- rename the API to stable
- leave an alias in place for now for the `__experimental` API so it continues to function until the end of the deprecation period

That would seem to cause the least impact. Given how prevalent the usage of this API is I agree we should treat it as a special case.

> I don't advise against deprecations for all experimental APIs as it would create a lot of bloat in our code base for no big added value for third-party developers.

Agreed. But as you say in certain cases it might be necessary.

**Comment by youknowriad** [comment]
On a second look to the wp-directory results. I think most of the results are false positives:

 - MailPoet is bundling Gutenberg (so includes its code)
 - Gutenberg (ok)
 - Pretty Links (it's a map file so not actual code)
 - Kubio and Email Marketting suggestions: I think it's not used at all (probably a copy/paste) or they're bundling a Gutenberg component as well (to confirm)

Anyway, I don't really think this API is that impactful (I don't see it as an important one for third-parties either). It's probably better if we just contact these plugins to confirm that it's false positives. We've done this in the past. cc @annezazu

**Comment by getdave** [comment]
FYI I think some of these instances could be based on using https://github.com/Automattic/isolated-block-editor/ as a base. 

I've contacted them and they've raised https://github.com/Automattic/isolated-block-editor/pull/135.

If we're not concerned then an alternative approach could be:

- promote the API without adding a deprecation
- ship in upcoming Gutenberg release
- prepare a followup PR ready to go (we could ship in a point release) just in case there are knock on effects reported

I wonder if @annezazu can reach out a bit and help identify the potential impact?

The other side of the coin is, if there is a level of uncertainty about the impact, perhaps we should just add the deprecation anyway. If we're worried about bloat, we can just tidy up after ourselves once the deprecation period has ended [as we did recently with Navigation Areas](https://github.com/WordPress/gutenberg/pull/40645).

**Comment by annezazu** [comment]
I'm on it :) Contacted the following folks today and linked to this issue for good measure: 
- Pretty Links
- Email Marketing Automation, Email Newsletter and CRM Plugin for WordPress by FluentCRM
- Kubio Page Builder

I'll loop back if I hear from them to confirm but, for now, consider proactive outreach done.

**Review by adamziel — APPROVED** [review]
I'm approving since it works and is consistent with Gutenberg __experimental policy. That's separate from handling the communication. A quick [WP directory](https://wpdirectory.net/search/01G2WFAAQBQQCJ5W6PEQP519ZW) search by `__experimentalFetchLinkSuggestions` revealed the following plugins:

- MailPoet – emails and newsletters in WordPress
- Gutenberg
- Pretty Links – Link Management, Branding, Tracking & Sharing Plugin
- Email Marketing Automation, Email Newsletter and CRM Plugin for WordPress by FluentCRM
- Kubio Page Builder
- Friends
- Gartentechnik.com
- Booking Weir

**Comment by annezazu** [comment]
Quick note for transparency that MailPoet is part of Automattic where I work (along with you) and they too have been contacted by @getdave previously. I left them off my above list for that reason but it feels important to clarify that they have been reached out to.

**Comment by getdave** [comment]
@youknowriad @adamziel I'm going to [flag this once more in Core Editor chat](https://make.wordpress.org/core/2022/05/16/editor-chat-agenda-18th-may-2022/#comment-43031) this week and then look to merge.

**Comment by getdave** [comment]
Here is where I flagged it again

https://wordpress.slack.com/archives/C02QB2JS7/p1652884611338199

Unless anyone else objects or ✋ then I will merge this before the end of the week.

**Comment by peterwilsoncc** [comment]
I share @Clorith's concern.

A sudden, hard deprecation of this in WordPress will break a lot of sites and the WordPress commitment to maintain backward compatibility.

As the conversation about making the block-editor less experimental is ongoing, I think it would be provocative to hard deprecate prior to the resolution.

Given the outreach to larger plugins in the plugin directory, of course the usage in the public code is going down but that's not indicative of custom themes, plugins, etc.

**Review by peterwilsoncc — CHANGES_REQUESTED** [review]
Marking this with a change request per my comment above. 

I've created a PR with said changes: #41183.

**Comment by getdave** [comment]
Closing this one.

**Pull request closed without merging**